doc/runtime_memory.rst

```diff
@@ -116,14 +116,104 @@ by both the host and the device. *Coarse-grain* SVM requires that
+    for dealing with this. In particular, all means for allocating SVM
+    allow specifying a :class:`CommandQueue`, so that deallocation
+    is enqueued and performed after previously-enqueued operations
+    have completed.
+
 SVM requires OpenCL 2.0.
 
+.. _opaque-svm:
+
+Opaque and "Wrapped-:mod:`numpy`" Styles of Referencing SVM
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When trying to pass SVM pointers to functionality in :mod:`pyopencl`,
+two styles are supported:
+
+- First, the opaque style. This style most closely resembles
+  :class:`Buffer`-based allocation available in OpenCL 1.x.
+  SVM pointers are held in opaque "handle" objects such as :class:`SVMAllocation`.
+
+- Second, the wrapped-:mod:`numpy` style. In this case, a :class:`numpy.ndarray`
+  (or another object implementing the  :c:func:`Python buffer protocol
+  <PyObject_GetBuffer>`) serves as the reference to an area of SVM.
+  This style permits using memory areas with :mod:`pyopencl`'s SVM
+  interfaces even if they were allocated outside of :mod:`pyopencl`.
+
+  Since passing a :class:`numpy.ndarray` (or another type of object obeying the
+  buffer interface) already has existing semantics in most settings in
+  :mod:`pyopencl` (such as when passin arguments to a kernel or calling
+  :func:`enqueue_copy`), there exists a wrapper object, :class:`SVM`, that may
+  be "wrapped around" these objects to mark them as SVM.
+
+The commonality between the two styles is that both ultimately implement
+the :class:`SVMPointer` interface, which :mod:`pyopencl` uses to obtain
+the actual SVM pointer.
+
+Note that it is easily possible to obtain a :class:`numpy.ndarray` view of SVM
+areas held in the opaque style, see :attr:`SVMPointer.buf`, permitting
+transitions from opaque to wrapped-:mod:`numpy` style. The opposite transition
+(from wrapped-:mod:`numpy` to opaque) is not necessarily straightforward,
+as it would require "fishing" the opaque SVM handle out of a chain of
+:attr:`numpy.ndarray.base` attributes (or similar, depending on
+the acutal object serving as the main SVM reference).
+
+See :ref:`numpy-svm-helpers` for helper functions that ease setting up the
+wrapped-:mod:`numpy` structure.
+
+Wrapped-:mod:`numpy` SVM tends to be a good fit for fine-grain SVM because of
+the ease of direct host-side access, but the creation of the nested structure
+that makes this possible is associated with a certain amount of cost.
+
+By comparison, opaque SVM access tends to be a good fit for coarse-grain
+SVM, because direct host access is not possible without mapping the array
+anyway, and it has lower setup cost. It is of course entirely possible to use
+opaque SVM access with fine-grain SVM.
+
+While all types of SVM can be used as the memory backing
+:class:`pyopencl.array.Array` objects, ensuring that new arrays returned
+by array operations (e.g. arithmetic) also use SVM is easiest to accomplish
+by passing an SVM
```

Comment:
The phrase is unfinished

Reply:
https://github.com/inducer/pyopencl/compare/b425777aec6de781501ba259c2fd376877b81f73..1d309eff22ce360b31f073d36ef90f4c85f02a92